apps/server/src/modules/files-storage/client/s3-client.adapter.ts

```diff
@@ -183,6 +186,50 @@ export class S3ClientAdapter implements IStorageClient {
 		return this.client.send(req);
 	}
 
+	public async list(prefix: string, maxKeys = Infinity) {
+		this.logger.log({ action: 'list', params: { prefix, bucket: this.config.bucket } });
+
+		try {
+			let files: string[] = [];
+			let ret: ListObjectsV2CommandOutput | undefined;
+
+			do {
+				const req = new ListObjectsV2Command({
+					Bucket: this.config.bucket,
+					Prefix: prefix,
+					ContinuationToken: ret?.NextContinuationToken,
+					MaxKeys: Math.min(maxKeys - files.length, 1000),
+				});
+
+				// Iterations are dependent on each other
+				// eslint-disable-next-line no-await-in-loop
+				ret = await this.client.send(req);
+
+				const returnedFiles =
+					ret?.Contents?.filter((o) => o.Key)
+						.map((o) => o.Key as string) // Can not be undefined because of filter above
+						.map((key) => key.substring(prefix.length)) ?? [];
+
+				files = files.concat(returnedFiles);
+			} while (ret?.IsTruncated && ret.NextContinuationToken && files.length < maxKeys);
```

Comment:
The docs say NextContinuationToken is always sent when IsTruncated=true. So you don't have to check that separately.  https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/listobjectsv2commandoutput.html#nextcontinuationtoken-4